### ISSUE TITLE
formatlist iteration count or interpolation range function

### ISSUE THREAD

**Issue by jasonkeene** [hashicorp/terraform#2769]
I want to be able to do a formatlist that includes a count of the iteration over the list:

```
formatlist("node-%d=http://%s:2380", COUNT, openstack_compute_floatingip_v2.demo.*.address)
```

There are two solutions. Have a custom printf format verb for the current iteration:

```
formatlist("node-%i=http://%s:2380", openstack_compute_floatingip_v2.demo.*.address)
```

It could easily be checked for here: https://github.com/hashicorp/terraform/blob/a81cad3672450151dd387a950e282d433ec4839d/config/interpolate_funcs.go#L164-L174

Or just provide a generic range function that will produce a list of integers:

```
formatlist("node-%d=http://%s:2380", range(length(openstack_compute_floatingip_v2.demo)), openstack_compute_floatingip_v2.demo.*.address)
```

I don't mind implementing this since I need it. I'd prefer the former.

**Comment by jasonkeene** [comment]
Note, count.index wouldn't work in my case since it's tied to the resource (in this case the instance) and this string needs to be created for each instance.

**Comment by jasonkeene** [comment]
Another solution I thought of is to provide the count index on the resource like this:

```
formatlist("node-%d=http://%s:2380", openstack_compute_floatingip_v2.demo.*.index, openstack_compute_floatingip_v2.demo.*.address)
```

It would be nice to have this on every resource that does count.

**Comment by jasonkeene** [comment]
@phinze gave me a workaround for this:

```
resource "template_file" "address-lines" {
    filename = "./address-line"
    count = "${var.size}"
    vars = {
        machine_num = "${count.index+1}"
        ip_address = "${element(openstack_compute_floatingip_v2.kuberdemo.*.address, count.index)}"
    }
}

resource "template_file" "user-data" {
    filename = "./user-data"
    count = "${var.size}"
    vars = {
        ...
        initial_cluster = "${join(", ", template_file.address-lines.*.rendered)}"
    }
}
```

I ran into the issue that my template data in ./address-line had a \n at the end due to vim that I had to remove. Obviously this is just a work around.

**Comment by jjn2009** [comment]
The problem I am having is that my user_data and my virtual machines cannot be dependent on each other, thus I cannot output a config line into my user data which reflects each machine I am going to create. If I had a range function then I could create a list derived from my `num_masters` variable (for example) which is passed to formatlist, which then constructs a string that contains each DNS name of the nodes I am going to deploy (given some canonicalized DNS scheme). `num_masters` would also be passed to `count` in my `digitalocean_droplet` for virtual machine creation and `digitalocean_record` for subdomain creation.

to illustrate further here is one of your examples you provide:
```
variable "instance_ips" {
  default = {
    "0" = "10.11.12.100"
    "1" = "10.11.12.101"
    "2" = "10.11.12.102"
  }
}

resource "aws_instance" "app" {
  count = "3"
  private_ip = "${lookup(var.instance_ips, count.index)}"
  # ...
}
```
In a somewhat different and simpler scenario than my own as described above I would ask: how can I programmatically create each ip in the instance ips subnet? this would require that the instance ips variable contains every ip in that range.

**Comment by vallard** [comment]
any updates on this or was another way to do this implemented?  Thanks!

**Comment by vallard** [comment]
For my workaround, using ```openstack_compute_instance_v2``` resources, when I created them I added a meta variable: 
```
resource "openstack_compute_instance_v2" "kube-worker" {
  name = "${var.worker_name}${format(var.count_format, count.index+1) }"
  count = "${var.worker_count}"
  image_name = "${var.kube_image}"
  flavor_name = "${var.kube_flavor}"
  metadata {
    kube_role = "worker"
    worker_number = "${count.index}"
  }
  network {
    name = "${var.network}"
  }
  key_pair = "${var.key_pair}"
  security_groups = ["${var.security_group}"]
}
```
Then when I needed to merge the lists I used the metadata.worker_number variable in formatlist: 
```
static_routes = "${join("\n", formatlist("up route add -net %s.%s.0/24 gw %s dev %s", var.cluster_nets_prefix, openstack_compute_instance_v2.kube-worker.*.metadata.worker_number,openstack_compute_instance_v2.kube-worker.*.access_ip_v4, var.if_dev))}"
```
In this case the *.metadata.worker_number provided the index I needed.

**Comment by blalor** [comment]
I stumbled upon the template_file solution myself, but it would really be nice to have a first-class `range` function, or something like `formatlist`.

**Comment by bloo** [comment]
@blalor ran into this as well - seems like a common use case

**Comment by r1ckr** [comment]
Any progress on the range function?

**Comment by apparentlymart** [comment]
Hi all! Sorry for the long silence here.

Our current project is integrating an improved version of the configuration language parser and interpreter that addresses some design flaws and adds some new features. Once we have completed our current work it should be possible to address the given use-case using the new "`for` expression" construct:

```hcl
# NOT YET IMPLEMENTED; details may change before final release

formatlist(
  "node-%d=http://%s:2380",
  [for i, v in openstack_compute_floatingip_v2.demo.*.address: i],
  openstack_compute_floatingip_v2.demo.*.address
)
```

Alternatively, this expression can be turned "inside-out", which seems more readable/understandable:

```hcl
# NOT YET IMPLEMENTED; details may change before final release

[
  for i, v in openstack_compute_floatingip_v2.demo.*.address:
  format("node-%d=http://%s:2380", i, v)
]
```

Integrating the improved implementation to get these new features is the current focus for the Terraform team at HashiCorp, and will be included in a forthcoming major release.

It is likely that a `range` function will still be added for other reasons, but I think it will be preferable to use the "`for` expression" construct for this sort of use-case where something needs to be evaluated for each index/value pair.

**Comment by apparentlymart** [comment]
We didn't include a build of the openstack provider in the subset of providers bundled with Terraform v0.12-alpha1, but I wrote a small configuration using the null provider to verify the behavior I mentioned in my last comment on that release:

```hcl

resource "null_resource" "example" {
  count = 4

  triggers = {
    address = cidrhost("10.1.0.0/16", count.index)
  }
}

output "example" {
  value = [
    for idx, inst in null_resource.example:
    format("node-%d=http://%s:2380", idx, inst.triggers["address"])
  ]
}
```

```
$ terraform output
example = [
  "node-0=http://10.1.0.0:2380",
  "node-1=http://10.1.0.1:2380",
  "node-2=http://10.1.0.2:2380",
  "node-3=http://10.1.0.3:2380",
]
```

This'll also be included in v0.12.0 final once we finish up the remaining issues and get through the beta/release-candidate cycle.
